### ISSUE TITLE
Decide of an API and use it

### ISSUE THREAD

**Issue by echebbi** [gemoc/ale-lang#123]
## Motivation

Some tasks are repeated often across the code but the current Java API is sometimes lacking (thus encouraging code duplication), sometimes cumbersome (thus arming code's readability) or even not documented enough.

I believe that it would save a great amount of time to have all of them clearly defined.

I hence propose to:

1. Identify those tasks
2. Define a user-friendly API to fulfill them
3. Document this API in ALE's website

## Proposition

Below are the tasks that come to my mind right now and the corresponding API.

| Task                                       | API                          | Documented?                             | User friendly?                                                                                                                                                                        |
| ------------------------------------------ | ---------------------------- | --------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Create a new ALE project                      | [WorkspaceAleProject](https://github.com/gemoc/ale-lang/blob/321725ff7f91291d152a55d31b5b5c802a0908f2/plugins/org.eclipse.emf.ecoretools.ale.ide.ui/src/org/eclipse/emf/ecoretools/ale/ide/ui/project/WorkspaceAleProject.java#L88)               | <ul align="center"><li>- [ ] </li></ul> | <ul><li>- [x] [Takes a lot of arguments](https://github.com/gemoc/ale-lang/blob/321725ff7f91291d152a55d31b5b5c802a0908f2/plugins/org.eclipse.emf.ecoretools.ale.ide.ui/src/org/eclipse/emf/ecoretools/ale/ide/ui/project/WorkspaceAleProject.java#L498) but OK</li></ul>                                                                                                                                    |
| Execute an ALE script                      | [ALEInterpreter](https://github.com/gemoc/ale-lang/blob/321725ff7f91291d152a55d31b5b5c802a0908f2/plugins/org.eclipse.emf.ecoretools.ale/src/org/eclipse/emf/ecoretools/ale/ALEInterpreter.java#L64)               | <ul align="center"><li>- [x] </li></ul> | <ul><li>- [x] Not straightforward but OK</li></ul>                                                                                                                                    |
| Retrieve environment's metamodel           | [AleProject](https://github.com/gemoc/ale-lang/blob/321725ff7f91291d152a55d31b5b5c802a0908f2/plugins/org.eclipse.emf.ecoretools.ale.ide/src/org/eclipse/emf/ecoretools/ale/ide/project/AleProject.java#L35) + [IAleEnvironment](https://github.com/gemoc/ale-lang/blob/321725ff7f91291d152a55d31b5b5c802a0908f2/plugins/org.eclipse.emf.ecoretools.ale.core/src/org/eclipse/emf/ecoretools/ale/core/interpreter/IAleEnvironment.java#L27) | <ul align="center"><li>- [ ] </li></ul> | <ul><li>- [ ] `IAleEnvironment` returns a list of URI as String instances; they almost always have to be transformed/loaded in the call site, which is definitely not handy</li></ul> |
| Retrieve environment's _.ale_ source files | [AleProject](https://github.com/gemoc/ale-lang/blob/321725ff7f91291d152a55d31b5b5c802a0908f2/plugins/org.eclipse.emf.ecoretools.ale.ide/src/org/eclipse/emf/ecoretools/ale/ide/project/AleProject.java#L35) + [IAleEnvironment](https://github.com/gemoc/ale-lang/blob/321725ff7f91291d152a55d31b5b5c802a0908f2/plugins/org.eclipse.emf.ecoretools.ale.core/src/org/eclipse/emf/ecoretools/ale/core/interpreter/IAleEnvironment.java#L27) | <ul align="center"><li>- [ ] </li></ul> | <ul><li>- [ ] `IAleEnvironment` returns a list of URI as String instances; they almost always have to be transformed/loaded in the call site, which is definitely not handy. Even worse, instead of URI [we may have absolute file paths](https://github.com/gemoc/ale-lang/blob/321725ff7f91291d152a55d31b5b5c802a0908f2/plugins/org.eclipse.emf.ecoretools.ale.ide/src/org/eclipse/emf/ecoretools/ale/ide/Normalized.java#L43)!</li></ul> |
| Manipulate classes declared in ALE | [AleProject](https://github.com/gemoc/ale-lang/blob/321725ff7f91291d152a55d31b5b5c802a0908f2/plugins/org.eclipse.emf.ecoretools.ale.ide/src/org/eclipse/emf/ecoretools/ale/ide/project/AleProject.java#L35) + [IAleEnvironment](https://github.com/gemoc/ale-lang/blob/321725ff7f91291d152a55d31b5b5c802a0908f2/plugins/org.eclipse.emf.ecoretools.ale.core/src/org/eclipse/emf/ecoretools/ale/core/interpreter/IAleEnvironment.java#L27) + [DslBuilder](https://github.com/gemoc/ale-lang/blob/321725ff7f91291d152a55d31b5b5c802a0908f2/plugins/org.eclipse.emf.ecoretools.ale.core/src/org/eclipse/emf/ecoretools/ale/core/parser/DslBuilder.java#L33) | <ul align="center"><li>- [ ] </li></ul> | <ul><li>- [ ] `DslBuilder` requires an AQL `IQueryEnvironment`. This is cumbersome anyway.</li></ul> |

**Comment by dvojtise** [comment]
Not sure it it not already covered by "Manipulate classes declared in ALE"But one recurrent activity is to read/set runtime data(for ex to create an UI or a GEMOC engine addon)With recurrent case for primitive typesMy best approach right now is to add an helper (for every RTD) written in ALE and call it, then convert the result...

**Comment by echebbi** [comment]
> Not sure it it not already covered by "Manipulate classes declared in ALE"But one recurrent activity is to read/set runtime data(for ex to create an UI or a GEMOC engine addon)

By "manipulate classes declared in ALE" I meant, for a given project/.dsl file/.ale file, having access to every [extended](https://github.com/gemoc/ale-lang/blob/321725ff7f91291d152a55d31b5b5c802a0908f2/plugins/org.eclipse.emf.ecoretools.ale.core/src-gen/org/eclipse/emf/ecoretools/ale/implementation/ExtendedClass.java#L36) and [runtime](https://github.com/gemoc/ale-lang/blob/321725ff7f91291d152a55d31b5b5c802a0908f2/plugins/org.eclipse.emf.ecoretools.ale.core/src-gen/org/eclipse/emf/ecoretools/ale/implementation/RuntimeClass.java#L25) class along with both their static (Ecore) and dynamic (ALE) attributes/operations.

You are rather talking about having access to the state of a runtime ALE program, such as variables' name and value, aren't you?

**Comment by echebbi** [comment]
I've updated my first post with code snippets.

**Comment by dvojtise** [comment]
Yes, (typical use : a view that shows the running state of the model)

and to my experience the current API need some documentation because it can easily lead to memory leaks if you don't dispose correctly the interpreter and the query environment (because it registers listeners)

**Comment by echebbi** [comment]
> Yes, (typical use : a view that shows the running state of the model)

I'm unsure about the kind of API you'd like. Since we give the `EObject` "root" to the ALE interpreter the model is already updated; from there we should be able to update the view with e.g. EMF notification, shouldn't we?

Is the issue to have access to the runtime data or rather to make it easily exploitable (e.g. turning `EString` into `String`)?

> and to my experience the current API need some documentation because it can easily lead to memory leaks if you don't dispose correctly the interpreter and the query environment (because it registers listeners)

Good point, I have updated the table.
